tools/storage_mon.c

```diff
@@ -72,65 +77,58 @@ static int open_device(const char *device, int verbose)
 	res = lseek(device_fd, seek_spot, SEEK_SET);
 	if (res < 0) {
 		fprintf(stderr, "Failed to seek %s: %s\n", device, strerror(errno));
-		close(device_fd);
-		return -1;
+		goto error;
 	}
 	if (verbose) {
 		printf("%s: reading from pos %ld\n", device, seek_spot);
 	}
-	return device_fd;
-}
-
-/* Check one device */
-static void *test_device(const char *device, int verbose, int inject_error_percent)
-{
-	int device_fd;
-	int sec_size = 0;
-	int res;
-	void *buffer;
-
-	if (verbose) {
-		printf("Testing device %s\n", device);
-	}
 
-	device_fd = open_device(device, verbose);
-	if (device_fd < 0) {
-		exit(-1);
-	}
+	if (flags & O_DIRECT) {
+		int sec_size = 0;
+		void *buffer;
 
 #ifdef __FreeBSD__
-	ioctl(device_fd, DIOCGSECTORSIZE, &sec_size);
+		res = ioctl(device_fd, DIOCGSECTORSIZE, &sec_size);
 #else
-	ioctl(device_fd, BLKSSZGET, &sec_size);
+		res = ioctl(device_fd, BLKSSZGET, &sec_size);
 #endif
-	if (sec_size == 0) {
-		fprintf(stderr, "Failed to stat %s: %s\n", device, strerror(errno));
-		goto error;
-	}
+		if (res < 0) {
+			fprintf(stderr, "Failed to stat %s: %s\n", device, strerror(errno));
+			goto error;
+		}
 
-	if (posix_memalign(&buffer, sysconf(_SC_PAGESIZE), sec_size) != 0) {
-		fprintf(stderr, "Failed to allocate aligned memory: %s\n", strerror(errno));
-		goto error;
-	}
+		if (posix_memalign(&buffer, sysconf(_SC_PAGESIZE), sec_size) != 0) {
+			fprintf(stderr, "Failed to allocate aligned memory: %s\n", strerror(errno));
```

Comment:
What I've read alignment should be more block-size. That makes sense thinking of a contiguous
block of memory cut into block-size chunks that you would like to be able to pass in any combination.
Usually it will probably be 512 bytes block-size vs. 4k pagesize and thus Ok.
Block-size of many disks being 4k nowadays I would expect that to become larger in future.
(or already in special cases like maybe raids, ssds with their large erase-block-size, shingled-write disks, ...)
max(block-size, pagesize) might be a way to satisfy everybody as much as possible.

Reply:
sbd is using block-size - at least since I've reworked that code-part a bit but iirc that was there since ever.

Reply:
aah - crap you aren't using aio here. So maybe forget what I've said.